### ISSUE TITLE
[Proposal] AreEquals/IsHomogeneous to test all elements of a sequence

### ISSUE THREAD

**Issue by Orace** [morelinq/MoreLINQ#698]
I propose for:
```C#
(bool isHomogeneous, T value) IsHomogeneous<T>(this IEnumerable<T> source, IEqualitityComparer<T> comparer = null);
```

The return tuple contains:

- `isHomogeneous`: true if all the elements in the source are equals relatively to `comparer` or `DefaultEqualitityComparer<T>`, false otherwise.
- `value`: the first value of the source if `isHomogeneous` is true, `default(T)` otherwise.

**Comment by Orace** [comment]
The overloads can be:

```C#
(bool? isHomogeneous, TSource value) IsHomogeneous<TSource>(this IEnumerable<TSource> source);
(bool? isHomogeneous, TSource  value) IsHomogeneous<TSource>(this IEnumerable<TSource> source, IEqualitityComparer<TSource> comparer);
(bool? isHomogeneous, TResult value) IsHomogeneous<TSource, TResult>(this IEnumerable<T> source, Func<TSource, TResult> selector);
(bool? isHomogeneous, TResult value) IsHomogeneous<TSource, TResult>(this IEnumerable<T> source, Func<TSource, TResult> selector, IEqualitityComparer<TResult> comparer);
```

**Comment by Orace** [comment]
An implementation of `IsHomogeneous` is available in #700:
https://github.com/morelinq/MoreLINQ/blob/365adcc40a94417299fc6101e361db959f04c6cd/MoreLinq/EquiInterleave.cs#L94-L110

**Comment by atifaziz** [comment]
Sounds like `Distinct` + [`TrySingle`](https://github.com/morelinq/MoreLINQ/tree/5ad0a6528a6d7708189518396026fd0c05df65c1#trysingle) (#540).

**Comment by Orace** [comment]
@atifaziz, it is but without O(1) memory usage.

**Comment by atifaziz** [comment]
Fair point but then thinking `HomogeneousOrEmpty` would be better? That is, return a sequence of one if all are the same or none. It's more idiomatic.

**Comment by atifaziz** [comment]
> @atifaziz, it is ~but with O(1) memory usage.~

Second thoughts?

**Comment by Orace** [comment]
> > @atifaziz, it is ~but with O(1) memory usage.~
> 
> Second thoughts?

Yep, distinc + TrySingle doesn't cost that much.

I'm ok with any name proposition.

**Comment by Orace** [comment]
`HomogeneousOrEmpty` will not distinct empty from non-homogeneous.

**Comment by atifaziz** [comment]
> Yep, distinc + TrySingle doesn't cost that much.

It doesn't cost if it's homogeneous but could otherwise.

> I'm ok with any name proposition.

First the signature then the name. Instead of asking a question or returning a Boolean, return a sequence of one if source is homogenous otherwise abort early and return an empty sequence (`yield break`). Joining with another sequence (e.g. `Zip`) would then have the right effect of returning a sequence or none.
 
> `HomogeneousOrEmpty` will not distinct empty from non-homogeneous.

Sorry, I don't understand.

**Comment by Orace** [comment]
> > `HomogeneousOrEmpty` will not distinct empty from non-homogeneous.
> 
> Sorry, I don't understand.

If the sequence is empty my proposal return null
If it's not homogeneous it's return false
`HomogeneousOrEmpty` return empty in both case, and you can't made a distinction between empty and not homogeneous

**Comment by atifaziz** [comment]
Fair point if you are after reduction without an error for an undefined state but then `Is…` is misleading because there is no Boolean result. You are looking for 3 states that define the _homogeneity_ of a source:

```c#
TResult Homogeneity<T, TResult>(
    this IEnumerable<TSource> source,
    TResult ambiguous, TResult heterogeneous, Func<T, TResult> homogeneous)
```

**Comment by Orace** [comment]
Indeed it's a 3 state. And this looks like the discussion about TrySingle.

The 3 delegates really add complexity, in my use case example it's obvious:
```C#
var (isHomogeneous, hasNext) = enumerators.Select(e => e.MoveNext()).IsHomogeneous(((bool?)null, default), (false, default), v  => (true, v)); 
 ```

That's why I prefer a new class or struct with a 3 state enum and the homogeneous value if any.

**Comment by Orace** [comment]
An other option is (optional via overloads) out parameters:

```C#
bool IsHomogeneous<T>(this IEnumerable<T> source, out bool hasValue, out T value);
```

But with the overloads to add a selector and a comparer it became a mess.

**Comment by atifaziz** [comment]
> Indeed it's a 3 state. And this looks like the discussion about TrySingle.

Yeah and now I am not happy about `TrySingle`. There's a sneaking problem underneath all this and I can't put my finger on it right now. I am tempted to remove `TrySingle` until I have had time to think about this.

> The 3 new parameters really add complexity, in my use case example it's obvious:
> 
> ```cs
> var (isHomogeneous, hasNext) = enumerators.Select(e => e.MoveNext()).IsHomogeneous(((bool?)null, default), (false, default), v  => (true, v)); 
> ```

The complexity comes from the choice of types.

> And it's source of bug since we can mess up with the two first parameters.

This is not always avoidable.

> That's why I prefer a new class or struct as return type with a 3 state enum and the homogeneous value if any.

This shouldn't be MoreLINQ's problem.

```c#
var (isHomogeneous, hasNext) = enumerators.Select(e => e.MoveNext()).IsHomogeneous(); 
```

This use case is not worth the trouble because it abuses LINQ for side-effects besides causing too many allocations.

**Comment by atifaziz** [comment]
> > The 3 new parameters really add complexity, in my use case example it's obvious:
> > ```cs
> > var (isHomogeneous, hasNext) = enumerators.Select(e => e.MoveNext()).IsHomogeneous(((bool?)null, default), (false, default), v  => (true, v)); 
> > ```
> 
> The complexity comes from the choice of types.

Let me expand on this. Suppose you introduce this in your project:

```c#
public enum TriBoolean { TriDefault, TriFalse, TriTrue }
```

then you can make it fairly readable:

```c#
// assume:
// using static TriBoolean;

var moves = from e in enumerators select e.MoveNext();
for (var stop = false; !stop;)
{
    switch (moves.Homogeneity(default, (TriFalse, default), moved => (TriTrue, moved)))
    {
        case (TriFalse, _):
            throw new InvalidOperationException("Input sequences are of different length.");
        case (TriTrue, true):
            foreach (var enumerator in enumerators) 
                yield return enumerator.Current;
            break;
        default:
            stop = true;
            break;
    }
}
```

Note that projection of moves can go outside the loop to save allocations.

> > That's why I prefer a new class or struct as return type with a 3 state enum and the homogeneous value if any.
> 
> This shouldn't be MoreLINQ's problem.

To expand on this too, if you provide a type, it will use generic names and won't make things necessarily more readable in the context of the actual operation. If you leave the choice up to the user then only the user is to blame or cheer for readability.

**Comment by Orace** [comment]
> > The 3 new parameters really add complexity, in my use case example it's obvious:
> > ```cs
> > var (isHomogeneous, hasNext) = enumerators.Select(e => e.MoveNext()).IsHomogeneous(((bool?)null, default), (false, default), v  => (true, v)); 
> > ```
> 
> The complexity comes from the choice of types.

I'm not agree with you, I need a type that can store any case (IsEmpty, IsHeteregoneous, IsHomogeneous with value T). I can't see something smaller than a 3 state (here `bool?`) and a value holder. 

> ```cs
> var (isHomogeneous, hasNext) = enumerators.Select(e => e.MoveNext()).IsHomogeneous(); 
> ```
> 
> This use case is not worth the trouble because it abuses LINQ for side-effects

Again not agree, this use case is pretty straightforward, I have an `IEnumerable<bool>` (it is `enumerators.Select(e => e.MoveNext())`) and I want to know it it's Empty, Heteregoneous or Homogeneous (with the value).

I _abuse_ linq since the enumeration stop early if an Heteregoneous case appear.
This kind of abuse are common (especially in linq):

    A() && B(); // do not evaluate B is A is false
    data.Select(selector).Where(validator).First(); // stop evaluation when a good case appear.

> [...] besides causing too many allocations.

Should I go down to a `for` loop ?

>Yeah and now I am not happy about TrySingle. There's a sneaking problem underneath all this and I can't put my finger on it right now. I am tempted to remove TrySingle until I have had time to think about this.

I am not happy about TrySingle neither.

There is a separation in the Linq method.

1 The _monad_ / _fluent_ ones: `Select`, `Where`, `Skip` who take an `IEnumerable` as input and give an `IEnumerable` at output.
2 The _consuming_ ones: `Count`, `First`, `Aggregate`, ...

`TrySingle` and the currently named `IsHomogeneous` are of the second type.
We expect from them to give an information about the source sequence.

For `Count`, `First`, `Aggregate` the type of the information is really simple : `int`, `T`, `TResult`, ...
For `TrySingle` and `IsHomogeneous` since they give more detailed information, if we want to put it in a returned object, the returned object became more complex. And the minimal form is a `(TriState, ValueHolder)`.

There is already complex object returned in Linq like [`IGrouping` in `GroupBy`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.groupby?view=netframework-4.8)

Like `Regexp.Match`, the returned type is complex.
And `Match` may became a `IEnumerable<char>` extension method, and we expect it to return a `Match` object.

For me your proposed solution for `TrySingle` is _functional programming oriented_:
 _we have 3 case, please provide expected behavior for each case_

An imperative form is:
_here your result, do what ever you want with it_

**Comment by Orace** [comment]
With a custom de-constructible `HomogeneityResult<T>` you can just write:

> 
> ```cs
> // assume:
> // using static HomogeneityType;
> 
> var moves = from e in enumerators select e.MoveNext();
> for (var stop = false; !stop;)
> {
>     switch (moves.Homogeneity())
>     {
>         case (Heterogeneous, _):
>             throw new InvalidOperationException("Input sequences are of different length.");
>         case (Homogeneous, true):
>             foreach (var enumerator in enumerators) 
>                 yield return enumerator.Current;
>             break;
>         default:
>             stop = true;
>             break;
>     }
> }
> ```

And add a `selector` or a `comparer` is painless.

> Note that projection of moves can go outside the loop to save allocations.

Good catch, I never do that since Resharper will grumble about multiple enumeration.

**Comment by leandromoh** [comment]
Does it is similar to this other [issue](https://github.com/morelinq/MoreLINQ/issues/266)?

**Comment by Orace** [comment]
> Does this PR is similar to this [other one](https://github.com/morelinq/MoreLINQ/issues/266)?

For the 'all items equal' part yes, but this proposed method also return the actual value (in the case of an homogeneous enumerable).
